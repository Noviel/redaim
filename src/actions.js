import { TRIGGER_UNKNOWN, TARGET_UNKNOWN, EXEC_UNKNOWN } from './constants';

export const createAction = 
  (trigger = TRIGGER_UNKNOWN) => 
  (target = TARGET_UNKNOWN) => 
  ({ exec = EXEC_UNKNOWN, payload = null, meta, ...rest } = {}) => 
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const aimFromAction = ({ type = '', payload = null, meta, ...rest } = {}) => {
  const [ 
    trigger = TRIGGER_UNKNOWN, 
    target = TARGET_UNKNOWN, 
    exec = EXEC_UNKNOWN 
  ] = type.split('/');

  return {
    ...rest,
    type: trigger || TRIGGER_UNKNOWN,
    payload,
    meta: {
      ...meta,
      target: target || TARGET_UNKNOWN,
      exec: exec || EXEC_UNKNOWN
    }
  };
};

export const wrapActions = fn => actions => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActions = actionCreatorWithTrigger => target => actions =>
  wrapActions(actionCreatorWithTrigger(target))(actions);
  
export const createActions = trigger => {
  const boundToTrigger = bindActions(createAction(trigger));

  return target => { 
    const boundToTarget = boundToTrigger(target);

    return actions => boundToTarget(actions);
  };
};

// reversed order of arguments
  
export const createActionR =
  ({ exec = EXEC_UNKNOWN, payload, meta, ...rest } = {}) =>
  (trigger = TRIGGER_UNKNOWN) =>
  (target = TARGET_UNKNOWN) =>
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const wrapActionsR = actions => fn => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActionsR = actions => {
  const wrappedActions = wrapActionsR(actions);

  return trigger => { 
    const actionAimer = createAction(trigger);

    return target => wrappedActions(actionAimer(target));
  };
};

export const createActionsR = actions => {
  const withAction = bindActionsR(actions);

  return trigger => {
    const withTrigger = withAction(trigger);

    return target => withTrigger(target);
  };
};
