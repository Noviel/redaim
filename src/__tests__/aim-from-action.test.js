import { aimFromAction } from '../actions';

describe('aimFromAction', () => {
  it('should create default aim if no action provided', () => {
    expect(aimFromAction()).toMatchSnapshot();
  });

  it('should translate simple action type to trigger', () => {
    expect(aimFromAction({ type: 'SIMPLE' })).toMatchSnapshot();
  });

  it('should translate action type with \'/\' to trigger and target', () => {
    expect(aimFromAction({ type: 'TRIGGER/TARGET' })).toMatchSnapshot();
  });

  it('should translate action type with two \'/\'s to trigger, target and exec', () => {
    expect(aimFromAction({ type: 'TRIGGER/TARGET/EXEC' })).toMatchSnapshot();
  });

  it('should merge other fields', () => {
    expect(aimFromAction({ 
      type: 'TRIGGER/TARGET/EXEC', 
      otherField: {
        data: 100
      }
    })).toMatchSnapshot();
  });

  it('should merge payload', () => {
    expect(aimFromAction({ 
      type: 'TRIGGER/TARGET/EXEC', 
      payload: { field1: 100, field2: { complex: 200 }}
    })).toMatchSnapshot();
  });

  it('should merge existing meta fields', () => {
    expect(aimFromAction({ 
      type: 'TRIGGER/TARGET/EXEC', 
      meta: { field1: 100, field2: { complex: 200 }}
    })).toMatchSnapshot();
  });

  it('should merge existing meta fields not overriding exec and target from type', () => {
    expect(aimFromAction({ 
      type: 'TRIGGER/TARGET/EXEC', 
      meta: { target: 'WRONG_TARGET', exec: 'WRONG_EXEC', other: 'OK' }
    })).toMatchSnapshot();
  });
});
