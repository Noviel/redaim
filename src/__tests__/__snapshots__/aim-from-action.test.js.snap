// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aimFromAction should create default aim if no action provided 1`] = `
Object {
  "meta": Object {
    "exec": "@@REDAIM-ACTION-UNKNOWN",
    "target": "@@REDAIM-TARGET-UNKNOWN",
  },
  "payload": null,
  "type": "@@REDAIM-TRIGGER-UNKNOWN",
}
`;

exports[`aimFromAction should merge existing meta fields 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC",
    "field1": 100,
    "field2": Object {
      "complex": 200,
    },
    "target": "TARGET",
  },
  "payload": null,
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should merge existing meta fields not overriding exec and target from type 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC",
    "other": "OK",
    "target": "TARGET",
  },
  "payload": null,
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should merge other fields 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC",
    "target": "TARGET",
  },
  "otherField": Object {
    "data": 100,
  },
  "payload": null,
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should merge payload 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC",
    "target": "TARGET",
  },
  "payload": Object {
    "field1": 100,
    "field2": Object {
      "complex": 200,
    },
  },
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should translate action type with '/' to trigger and target 1`] = `
Object {
  "meta": Object {
    "exec": "@@REDAIM-ACTION-UNKNOWN",
    "target": "TARGET",
  },
  "payload": null,
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should translate action type with two '/'s to trigger, target and exec 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC",
    "target": "TARGET",
  },
  "payload": null,
  "type": "TRIGGER",
}
`;

exports[`aimFromAction should translate simple action type to trigger 1`] = `
Object {
  "meta": Object {
    "exec": "@@REDAIM-ACTION-UNKNOWN",
    "target": "@@REDAIM-TARGET-UNKNOWN",
  },
  "payload": null,
  "type": "SIMPLE",
}
`;
